### ISSUE TITLE
Report other mouse buttons

### ISSUE THREAD

**Issue by remme123** [alacritty/alacritty#4754]
I've tried map mouse button X1 & X2 in vim, but nothing happened when I clicked. And double click also heppened nothing.

**Comment by remme123** [comment]
vim map command: map <X1Mouse> :echo '1'<cr>

**Comment by chrisduerr** [comment]
What are mouse buttons X1 and X2?

**Comment by remme123** [comment]
As described in the help of vim:
> The X1 and X2 buttons refer to the extra buttons found on some mice.  The
'Microsoft Explorer' mouse has these buttons available to the right thumb.
Currently X1 and X2 only work on Win32 and X11 environments.

**Comment by chrisduerr** [comment]
Are you aware of any other terminal emulator supporting this, or does this just work in gvim?

In theory supporting this in Alacritty would be possible and it currently does not correctly send these mouse buttons, but the way this is phrased with Win32 and X11 makes me question if it is supported by vim inside the terminal.

**Comment by remme123** [comment]
Yes, I've only seen it in gvim. It will be great if alacritty supports this feature.

**Comment by nixpulvis** [comment]
@remme123 I'm assuming you didn't mean to close this issue, reopening.

**Comment by chrisduerr** [comment]
It's possible that this cannot be supported in terminal vim, simply because vim doesn't support it. But that doesn't mean that Alacritty can't provide support for it.

**Comment by sophrosyne97** [comment]
@chrisduerr I am interested in closing this issue, I just have a couple of questions.

What exactly is meant by "Report other mouse buttons"? Does this just mean that we should allow a user to map to whatever extra mouse buttons they have, or does report mean something more like actually showing them that these buttons are available somehow? In on_mouse_press() match statement in input.rs I see a comment saying "Can't properly report more than three buttons..", would properly handling most extra mouse buttons by returning a value for each one here be a start towards this?

In looking into mapping extra mouse buttons I've briefly looked through the code and saw that you use a MouseButton enum with one of the variants being a tuple struct Other(u16). I am assuming this u16 is for referring to the particular button value? For example, using xev I know that the extra buttons on my mouse refer to button8 and button9; continuing off what I asked in the previous paragraph about input.rs, would MouseButton::Other(8) properly match to button8 somehow? I am not really sure how to properly read the correct extra mouse button and then handle it with the MouseButton enum.

One last thing is there a resource for seeing common names for these extra buttons? I know my two extra ones, like i've said before, are button8 and button9 but I know there are mouses that have many more. How do I handle mouses with far more buttons that I am not aware of so that people can bind to them anyway as well?

**Comment by chrisduerr** [comment]
> What exactly is meant by "Report other mouse buttons"? Does this just mean that we should allow a user to map to whatever extra mouse buttons they have, or does report mean something more like actually showing them that these buttons are available somehow?

Report means reporting it to terminal applications.

> would MouseButton::Other(8) properly match to button8 somehow?

That should be trivial to test yourself. Being able to do so is kind of necessary to implement this anyway.

**Comment by alex-courtis** [comment]
> Are you aware of any other terminal emulator supporting this, or does this just work in gvim?

st, with or without tmux.

> It's possible that this cannot be supported in terminal vim, simply because vim doesn't support it. But that doesn't mean that Alacritty can't provide support for it.

This does work in both terminal vim and neovim.

neovim scrolls up/down by a page like `^F` and `^B`.
```vim
" necessary
set mouse=a
" just to see what is going on
set number
```

vim scrolls 3 lines like `3^E` and `3^Y`.
```vim
" necessary
set ttymouse=sgr
set mouse=a
" just to see what is going on
set number
```

The fact that we need `sgr` might imply that it is not simple keystrokes being sent, but the actual mouse actions themselves. alacritty supports mouse actions e.g. dragging beautifully so this might be easier to implement than we think.

Further testing with st shows that those mouse events are not sent to stdin.

**Comment by alex-courtis** [comment]
Further testing with st shows that those mouse events are not sent to stdin.

Debugging st has revealed that it sends the following escape sequence prefixes in sgr mode:
Scroll wheel up: `\033[<64;`
Scroll wheel down: `\033[<65;`
Scroll wheel left: `\033[<66;`
Scroll wheel right: `\033[<67;`
Back: `\033[<68;`
Forward: `\033[<69;`
See [XTerm Control Sequences](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Other-buttons) for more information.

The bad news is that none of these appear able to be remapped via vim's `<XNMouse>`; the scroll funtionality described above remains unchanged. I may dig into this further.

The compromise solution could be to the ability to configure alacritty to send keystrokes on mouse buttons e.g. <F11> <F12>, which we could then map to our desired vim functionality.

**Comment by samusz** [comment]
May be remove the D-Easy tag as this seems quite involved with X and (neo)vim layers

**Comment by chrisduerr** [comment]
It's not.

**Comment by marianomarciello** [comment]
Hello everyone. I am trying to debug this issue. I have added the following mapping in my configuration file:
```yml
mouse_bindings:
-  {  mouse:  8,      action:  SpawnNewInstance }
```

Unfortunately I have only an extra button on my mouse (button 8). After changing the input.rs file a bit I am able to use the previous mouse binding.

Do you know some software that can emulate other mouse buttons ? I found xmodmap but it only support (Left, Right, Middle buttons and wheel)

**Comment by kchibisov** [comment]
You could open a PR if you want to get this actually solved.

**Comment by VishnuSanal** [comment]
hi, is this still open? I would like to work on this. would reviving #6957 do? I am a bit new to rust and looking forward to contributing. :)

**Comment by chrisduerr** [comment]
#6957 was closed because the author seemingly lost interest. So yes, if you're interested in picking this up then feel free to start working on it.

However please do not open a new PR unless the requested changes in the old PR have been made. If you're struggling and need help, you can ask questions in the old PR.
